tests/test_trainer.py

```diff
@@ -1690,3 +1692,85 @@ def hp_name(trial):
             trainer.hyperparameter_search(
                 direction="minimize", hp_space=hp_space, hp_name=hp_name, backend="sigopt", n_trials=4
             )
+
+
+@require_torch
+class TrainerOptimizerChoiceTest(unittest.TestCase):
+    def test_invalid_optimizer(self):
+        args = TrainingArguments(optim="bla", output_dir="None")
+        with self.assertRaises(ValueError):
+            Trainer.get_optimizercls_and_params(args)
+
+    def check_optim(self, args, mandatory_params, expected_cls):
+        """
+        Checks that the common case for an optimizer works.
+        """
+        actual_cls, optim_params = Trainer.get_optimizercls_and_params(args)
+        self.assertEqual(expected_cls, actual_cls)
+        self.assertIsNotNone(optim_params)
+
+        for p, v in mandatory_params.items():
+            self.assertTrue(p in optim_params)
+            actual_v = optim_params[p]
+            self.assertTrue(actual_v == v, f"Failed check for {p}. Expected {v}, but got {actual_v}.")
+
+    def test_adafactor(self):
+        from transformers.optimization import Adafactor
+
+        args = TrainingArguments(optim=OptimizerNames.ADAFACTOR.value, output_dir="None")
+
+        mandatory_params = {"scale_parameter": False, "relative_step": False}
+
+        self.check_optim(args, mandatory_params, Adafactor)
+
+    def test_adam_hf(self):
+        from transformers.optimization import AdamW
+
+        args = TrainingArguments(optim=OptimizerNames.ADAM_HF.value, output_dir="None", learning_rate=0.3)
+
+        mandatory_params = {
+            "betas": (args.adam_beta1, args.adam_beta2),
+            "eps": args.adam_epsilon,
+            "lr": args.learning_rate,
+        }
+
+        self.check_optim(args, mandatory_params, AdamW)
+
+    def test_adam_torch(self):
```

Comment:
This is almost identical to the previous test - use `parameterized` and have 2 tests with a single body?

This could include the real apex test as well, except it'd need to skip if `is_apex_available` is `False`.

```
from transformers.file_utils import is_apex_available
...
```